### DISCUSSION TITLE
Larger images despite VRAM size

### DISCUSSION THREAD

**Discussion by OldManGir42** [vladmandic/sdnext#1531] [Q&A] [answered]
Is there a good way to generate large images?

I'm running on windows on a 6900XT (16g) and any time I try to generate an image larger than about 368x720 the render drops and it gives me a runtime error: Could not allocate tensor with XXXXXXXXXXXX bytes. There is not enough GPU video memory available. 

Even trying to run the Hires fix at only 1x upscale throws the same error. Am I just stuck with small images based on my vram?

As a side note...once I generate my first image for a session my VRAM utilization stays maxed out until I close out of the command prompt window. Not sure if that's normal.

**Comment by vladmandic** [comment]
torch-directml does not implement autocast, so to avoid NaN issues all ops are in fp32 mode which uses double the memory than usual.
unfortunate, but it torch-directml is far from great. i hope amd releases rocm for windows soon, that would result in at least 4x performance and 2x memory reduction.

regarding vram utilization, again, thats how torch-directml works - it keeps internal memory pool. meaning that memory may be free for use for torch, but you cant see it from the outside. and there is no api to check the memory other than very basics (cuda is much more detailed).

**Comment by brknsoul** [comment, marked as answer]
A couple of tips that @aptronymist and I have played around with. With my 6600 XT 8Gb, I can gen 1024x1024 images comfortably.

This may or may not work for you;

Run `webui.bat --use-directml --experimental`

Then use these settings in Compute Settings;

![image](https://github.com/vladmandic/automatic/assets/1969381/4a387c86-43b5-44ba-a54d-4c1e10d7102a)

and in Optimizations, use the Sub-quadratic Cross-attention optimization method.

Then close and restart the server (ensuring you use the same command above).

What we've found is by _not_ Upcasting the Cross Attention Layer, this results in no black images.

**Reply by Aptronymist** [reply]
*ahem* I knew this all along. Brknsoul was playing. I'm proud though! 😆

**Reply by OldManGir42** [reply]
Thanks a ton! That definitely worked and made playing around with it a ton more interesting with some very...alien...results

**Reply by brknsoul** [reply]
Try genning some landscapes/cityscapes with something like 512x1536. ;-)

**Reply by OldManGir42** [reply]
I've generated some people that looked like they belonged in the next Human Centipede movie

**Reply by brknsoul** [reply]
When genning people, you should use 512x768 and then use Hires. fix.

SD has the propensity to 'fill in' extra space, sometimes with clones/extra people, sometimes with extra body parts ;-)

**Reply by OldManGir42** [reply]
I've definitely been noticing that. Is there an efficient way to do some outpainting through this UI? I'm trying to essentially take generated characters (at the reccomended 512x768) and add to the background to make a full 4k wallpaper...

**Reply by FrakerKill** [reply]
I try this and yes you can generate pics 1024x1024 but size is about 700KB before with another webui of lshqqytiger, I generated 840x840 pics of 1-1.3MB. Which could be the optimisation for that?

I have RX6600 8GB too

**Reply by vladmandic** [reply]
well, direct-ml support here is written by lshqqytiger and afaik, this codebase is newer than his fork - if there are functional differences, ping him directly.

**Comment by Aptronymist** [comment]
> A couple of tips that @Aptronymist and I have played around with

Say what?! I told you exactly what to do from the start! 😛
Don't make me come over there!

**Comment by Threnos** [comment]
I highly recommend reading [this guide about Ultimate SD Upscale](https://civitai.com/models/59811/4k-resolution-upscale-8x-controlnet-tile-resample-in-depth-with-resources). It saves much more from the original image while adding details, than usual hires fix or img2img, and, most importantly, easier on vram requirements as processing is done in tiles.
And I found [NMKD Siax 200k upscaler](https://icedrive.net/s/43GNBihZyi) to be superior to any built-in one.
